upb/msg_internal.h

```diff
@@ -131,7 +132,11 @@ UPB_INLINE bool _upb_hasbit(const upb_msg *msg, size_t idx) {
 }
 
 UPB_INLINE void _upb_sethas(const upb_msg *msg, size_t idx) {
-  (*UPB_PTR_AT(msg, idx / 8, char)) |= (char)(1 << (idx % 8));
+  if (UPB_LIKELY(idx < 32)) {
+    (*UPB_PTR_AT(msg, 0, uint32_t)) |= (1UL << idx);
+  } else {
+    (*UPB_PTR_AT(msg, idx / 8, char)) |= (char)(1 << (idx % 8));
+  }
```

Comment:
I am skeptical of this micro optimization. I think it would be better to pull this into a separate change.  micro optimizations like this often look good in benchmarks but introduce more code which then polutes your icache in real world systems

Reply:
`__builtin_expect()` (aka `UPB_LIKELY()`) will make the likely case into straight-line code, and put the unlikely case far away in memory that will likely never hit the cache unless the unexpected case is hit. https://godbolt.org/z/P1fodxfxd

Also the size profiling results show that this particular optimization only grows the code by 16 or 32 bytes total (this optimization was pulled into `upb_msg_set()` and `upb_msg_mutable()`.

I can move this to a separate change, but I think it is a well-grounded optimization.

Reply:
I would add though that I share your concern about icache in general, and upb is very size-sensitive in general, so I appreciate the attention to this. :)

Reply:
Benchmarks at the moment seem to be looking better without it, so it looks like it was a good call, thanks.

Reply:
the / and the % are likely a single instruction.  Even though they are an expensive instruction, it is still just one